vowpalwabbit/offset_tree_internal.cc

```diff
@@ -0,0 +1,138 @@
+#include "offset_tree_internal.h"
+#include "vw_exception.h"
+
+namespace offset_tree {
+  tree_node::tree_node(uint32_t i, uint32_t l, uint32_t r, bool lf)
+    : id(i), left_id(l), right_id(r), is_leaf(lf) {}
+
+  bool tree_node::operator==(const tree_node& rhs) const {
+    if (this == &rhs) return true;
+    return (id == rhs.id && left_id == rhs.left_id && right_id == rhs.right_id && is_leaf == rhs.is_leaf);
+  }
+
+  bool tree_node::operator!=(const tree_node& rhs) const { return !(*this == rhs); }
+
+  min_depth_binary_tree::min_depth_binary_tree()
+    : root_idx{0}, _num_leaf_nodes{0}, _initialized{false} {}
+
+  void min_depth_binary_tree::build_tree(uint32_t num_nodes) {
+    // Sanity checks
+    if (_initialized) {
+      if (num_nodes != _num_leaf_nodes) {
+        THROW("Tree already initialized.  New leaf node count (" << num_nodes << ") does not equal current value. (" << _num_leaf_nodes << ")");
+      }
+      return;
+    }
+
+    _num_leaf_nodes = num_nodes;
+    // deal with degenerate cases of 0 and 1 actions
+    if(_num_leaf_nodes == 0) {
+      _initialized = true;
+      return;
+    }
+
+    try {
+      nodes.reserve(2 * _num_leaf_nodes - 1);
+      std::vector<uint32_t> tournaments;
```

Comment:
Can you explain the use of the work tournament?

Reply:
This is probably a leftover from the ECT code although it is valid to think of the offset tree as conducting a single elimination tournament over actions.